File: workshop/example/python/template.py
from parser import OneOf, character, greedy, greedy1, pack, satisfy, token

def use(templateResult):
    return templateResult[0][0]

def template():
    return greedy(literal().orElse(substitution()))\
        .map(lambda result: Concatenation(result))

def literal():
    return greedy1(escapedCharacter().orElse(notAnOpeningBracket()))\
        .map(lambda result: ''.join(result))\
        .map(lambda verbatim: Literal(verbatim))

def escapedCharacter():
    return OneOf([
        token('\\').map(lambda _: '\\'),
        token('\\{').map(lambda _: '{'),
    ])

def notAnOpeningBracket():
    return satisfy(lambda character: character != '{')

def notAnClosingBracket():
    return satisfy(lambda character: character != '}')

def substitution():
    return pack(openingBracket(), identifier(), closingBracket())\
        .map(lambda identifier: Variable(identifier))

def identifier():
    return greedy1(notAnClosingBracket())\
        .map(lambda result: ''.join(result))\

def openingBracket():
    return character('{')

def closingBracket():
    return character('}')

class Template:
    def __call__(self, context):
        raise NotImplementedError()

class ValueMissingFromContext(Exception):
    pass

class Literal(Template):
    def __init__(self, literal):
        self.literal = literal

    def __call__(self, context):
        return self.literal

class Variable(Template):
    def __init__(self, name):
        self.name = name

    def __call__(self, context):
        if self.name in context:
            return context[self.name]
        else:
            raise ValueMissingFromContext(f'expected \'{self.name}\' to be in context')

class Concatenation(Template):
    def __init__(self, templates):
        self.templates = templates
    
    def __call__(self, context):
        return ''.join([template(context) for template in self.templates])

if __name__ == '__main__':
    context = { 'subject': 'World' }
    assert Literal('Hello, World!')(context) == 'Hello, World!'
    assert Variable('subject')(context) == 'World'

    try:
        Variable('missing from context')(context)
        assert False
    except ValueMissingFromContext as e:
        assert True
    
    assert Concatenation([])(context) == ''
    assert Concatenation([Literal('Hello, '), Variable('subject'), Literal('!')])(context) == 'Hello, World!'

    use(literal()('Hello, World!'))(context) == 'Hello, World!'
    use(literal()('\\\\'))(context) == '\\'
    use(literal()('\\{'))(context) == '{'

    use(substitution()('{subject}'))(context) == 'World'

    use(template()('Hello, {subject}!'))(context) == 'Hello, World!'

File: workshop/example/node/template.js
const assert = require('assert').strict
const Parser = require('./parser')
const Util = require('./util')

function use(result) {
    return function(context) {
        return result[0][0].evaluate(context)
    }
}

class Literal {
    constructor(literal) {
        this.literal = literal
    }

    evaluate(context) {
        return this.literal
    }
}

class Variable {
    constructor(name) {
        this.name = name
    }

    evaluate(context) {
        if (this.name in context) {
            return context[this.name]
        } else {
            throw "expected '" + this.name + "' to be in context"
        }
    }
}

class Concatenation {
    constructor(templates) {
        this.templates = templates
    }

    evaluate(context) {
        return this.templates
            .map(function (template) { return template.evaluate(context) })
            .join('')
    }
}

function template() {
    return Parser.just(Parser.map(Util.greedy(Parser.or(literal(), substitution())), function (templates) {
        return new Concatenation(templates)
    }))
}

function literal() {
    return Parser.map(Util.greedy1(Parser.or(
        escapedCharacter(),
        notAnOpeningBracket()
    )), function (result) { return new Literal(result.join('')) })
}

function escapedCharacter() {
    return Util.oneOf([
        Parser.map(Parser.token("\\\\"), function () { return "\\" }),
        Parser.map(Parser.token("\\{"), function () { return "{" })
    ])
}

function notAnOpeningBracket() {
    return Parser.satisfy(function (c) { return c != '{' })
}

function notAnClosingBracket() {
    return Parser.satisfy(function (c) { return c != '}' })
}

function substitution() {
    return Parser.map(Util.pack(
        openingBracket(),
        identifier(),
        closingBracket()
    ), function (identifier) { return new Variable(identifier) })
}

function identifier() {
    return Parser.map(
        Util.greedy1(notAnClosingBracket()),
        function (result) { return result.join('') }
    )
}

function openingBracket() {
    return Parser.character('{')
}

function closingBracket() {
    return Parser.character('}')
}

context = { 'subject': 'World' }
assert.equal((new Literal('Hello, World!')).evaluate(context), 'Hello, World!')
assert.equal((new Variable('subject')).evaluate(context), 'World')

try {
    (new Variable('missing from context')).evaluate(context)
    assert.fail
} catch (e) {
    assert.equal(e, "expected 'missing from context' to be in context")
}

assert.equal(use(literal()("Hello, World!"))(context), "Hello, World!")
assert.equal(use(literal()("\\\\"))(context), "\\")
assert.equal(use(literal()("\\{"))(context), "{")

assert.equal(use(substitution()("{subject}"))(context), "World")

assert.equal(use(template()("Hello, {subject}!"))(context), "Hello, World!")